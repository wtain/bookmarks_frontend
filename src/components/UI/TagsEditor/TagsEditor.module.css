
.container {
    border-style: inset;
    /* flex-wrap: wrap-reverse; */
    /* display: flex; */
    flex-flow: column wrap;
    /* width: 278px; */
    cursor: text;
}

.sp {
    /* background-color: lightpink; */
    /* display: flex; */
    /* float: left; */
    display: inline-block;
    width: auto;
}

.input {
    background-color: transparent;
    border: 0px;
    display: table-cell;
    width: 100%;
    color: white;
}

.input:focus {
    border: 0px;
    outline: none;
}

.tag {
    border-radius: 5px;
    margin: 1px;
    width: max-content;
    background-color: darkblue;
    display: inline-block;
    color: lightgray;
    padding: 1px;
    border-style: outset;
    border-color: black;
}

.clickable {
    cursor: pointer;
}

.close_button_wrapper {
    float: right;
}

.close_button {
    width: 20px;
    height: 20px;
    border-radius: 10px;
    position: relative;
    right: 0px;
}