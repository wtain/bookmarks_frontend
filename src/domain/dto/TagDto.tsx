
import React from "react";

interface TagDto {
    name: string;
    id: string;
}

export default TagDto;