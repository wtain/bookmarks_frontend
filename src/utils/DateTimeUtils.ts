import moment from "moment";

export function parseDateTimeFromMongo(datetime: Date | undefined) {
  if (datetime === undefined) {
    return undefined;
  }
  return moment(datetime.toISOString()).fromNow();
}

export function addOneMonth(date: Date): Date {
  let result = new Date(date);
  result.setMonth(result.getMonth() + 1)
  return result;
}

export function firstDayOfMonth(date: Date): Date {
  let result = new Date(date);
  result.setDate(1)
  return result;
}

export function getDatePortion(date: Date): string {
  const dateCopy = new Date(date);
  dateCopy.setUTCHours(0);
  dateCopy.setMinutes(0);
  const momentForDate = moment(dateCopy);
  return momentForDate.format('YYYY-MM-DD');
}

export function datesEqual(date1: Date, date2: Date): boolean {
  return getDatePortion(date1) === getDatePortion(date2);
}