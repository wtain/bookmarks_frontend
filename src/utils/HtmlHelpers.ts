import sanitizeHtml from "sanitize-html";

export function sanitizeText(text: string){
  return sanitizeHtml(text, {
    allowedTags: ['br', 'b', 'a', 'img']
  });
}