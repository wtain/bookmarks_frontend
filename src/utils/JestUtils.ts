
export function jestHacks() {
    // https://stackoverflow.com/questions/39830580/jest-test-fails-typeerror-window-matchmedia-is-not-a-function
    // https://github.com/facebook/create-react-app/issues/10126
    window.matchMedia = (query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })
}