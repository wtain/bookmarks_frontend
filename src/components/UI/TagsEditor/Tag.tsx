
import React, { useState } from "react";
import TagDto from "../../../domain/dto/TagDto";
import cl from './TagsEditor.module.css'

interface Props {
    tag: TagDto;
    onDelete?: () => void;
    onClick?: () => void;
}

const Tag: React.FC<Props> = (props: Props) => {

    const [showClose, setShowClose] = useState(false)
    const [clickablePointer, setClickablePointer] = useState(false);

    const clickable = cl.clickable;

    const styles = [cl.tag];

    if (clickablePointer) {
        styles.push(clickable);
    }

    return (
        <span className={styles.join(" ")} 
            onClick={() => {
                if (props.onClick) {
                    props.onClick()
                }
            }}
            onMouseEnter={() => {
                setShowClose(true);
                if (props.onClick) {
                    setClickablePointer(true);
                }
            }}
            onMouseLeave={() => {
                setShowClose(false);
                setClickablePointer(false);
            }}>
            {props.tag.name}
            {
                (showClose && props.onDelete) ? 
                    <div className={cl.close_button_wrapper}>
                        <button className={cl.close_button} onClick={() => {
                                if (props.onDelete !== null && props.onDelete !== undefined)
                                    props.onDelete!()
                            }}>
                            x
                        </button> 
                    </div>
                    : <></>
            }
        </span>
    )
}

export default Tag;