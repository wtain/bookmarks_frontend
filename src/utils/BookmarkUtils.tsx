
import React from "react";
import BookmarkDto from "../domain/dto/BookmarkDto";
import {v4 as uuidv4} from 'uuid';
import TagDto from "../domain/dto/TagDto";

const createBookmark = (summary: string, contents: string, tags: TagDto[]): BookmarkDto => {
    return {
                summary, 
                contents, 
                id: uuidv4(), 
                created: new Date().toLocaleString(),
                tags,
                isDone: false
            }
}

const createNewTag = (): TagDto => {
    return { name: "", id: uuidv4() }
}

const BookmarkUtils = { createBookmark, createNewTag };

export default BookmarkUtils;