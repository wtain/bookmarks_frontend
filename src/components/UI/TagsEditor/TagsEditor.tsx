
import React, { useEffect } from "react";
import TagDto from "../../../domain/dto/TagDto";
import BookmarkUtils from "../../../utils/BookmarkUtils";
import Tag from "./Tag";
import cl from './TagsEditor.module.css'

interface Props {
    tags: TagDto[];
    onTagAdded: (tag: TagDto) => void;
    onDelete: (index: number) => void;
    containerClass?: string;
}

const TagsEditor: React.FC<Props> = (props: Props) => {

    const [currentTag, setCurrentTag] = React.useState(BookmarkUtils.createNewTag());
    const inputRef = React.useRef<null | HTMLInputElement>(null);

    useEffect(() => {
        setCurrentTag(BookmarkUtils.createNewTag());
    }, [])

    const onTagAdded = () => {
        let value = currentTag.name.trim()
        if (value.length > 0) {
            if (props.tags.find(t => t.name === value) === undefined) {
                props.onTagAdded({...currentTag, name: value});                
            }
            setCurrentTag(BookmarkUtils.createNewTag());
        }
    }

    const onCurrentTagChange = (v: string) => {
        let p = v.indexOf(" ")
        if (p !== -1) {
            let value = v.substring(0, p).trim();
            if (value.length > 0) {
                props.onTagAdded({...currentTag, name: value});
            }
            setCurrentTag({ ...BookmarkUtils.createNewTag(), name: v.substring(p+1) })
        } else {
            setCurrentTag({ ...currentTag, name: v })
        }
    }

    const containerClasses = [cl.container];
    if (props.containerClass) {
        containerClasses.push(props.containerClass);
    }

    return (
        <div className={containerClasses.join(" ")} onClick={(e) => {
            inputRef.current?.focus();
        }}>
            {
                props.tags.map((tag, i) => <Tag key={tag.id} tag={tag} onDelete={() => props.onDelete(i)} />)
            }
            <span className={cl.sp}>
                <input ref={inputRef}
                        value={currentTag.name} 
                        onChange={(e) => onCurrentTagChange(e.target.value)} 
                        className={cl.input} 
                        placeholder="Tags"
                        onKeyDown={(e) => {
                            if (e.code === "Space" || e.code === "Tab") {
                                if (currentTag.name.trim().length > 0) {
                                    onTagAdded();
                                    e.stopPropagation();
                                }
                            }
                        }} 
                        onBlur={(e) => {
                            if (currentTag.name.trim().length > 0) {
                                onTagAdded();
                            }
                        }}
                        />
            </span>
        </div>
    )
}

export default TagsEditor;