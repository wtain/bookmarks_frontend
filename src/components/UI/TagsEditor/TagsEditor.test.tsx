import renderer from 'react-test-renderer';
import TagDto from '../../../domain/dto/TagDto';
import TagsEditor from './TagsEditor';
import { AppProvider } from "@shopify/polaris";
import translations from "@shopify/polaris/locales/en.json";
import { jestHacks } from '../../../utils/JestUtils';

describe('TagsEditor tests', () => {
  beforeAll(() => {
    jestHacks();
  });

  it('TagsEditor renders correctly', () => {

    const tags: TagDto[] = [
      {
        name: "tag1",
        id: "tag1-id"
      },
      {
        name: "tag2",
        id: "tag2-id"
      },
      {
        name: "tag3",
        id: "tag3-id"
      }
    ]

    const tree = renderer
      .create(
        <AppProvider i18n={translations}>
          <TagsEditor tags={tags} />
        </AppProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});