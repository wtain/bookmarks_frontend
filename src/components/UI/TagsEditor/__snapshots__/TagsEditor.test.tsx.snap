// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsEditor tests TagsEditor renders correctly 1`] = `
Array [
  <div
    className="container"
    onClick={[Function]}
  >
    <span
      className="Polaris-Tag Polaris-Tag--removable"
    >
      <span
        className="Polaris-Tag__TagText"
        title="tag1"
      >
        tag1
      </span>
      <button
        aria-label="Remove tag1"
        className="Polaris-Tag__Button"
        disabled={false}
        onClick={[Function]}
        onMouseUp={[Function]}
        type="button"
      >
        <span
          className="Polaris-Icon"
        >
          <span
            className="Polaris-Text--root Polaris-Text--visuallyHidden"
          />
          <svg
            aria-hidden="true"
            className="Polaris-Icon__Svg"
            focusable="false"
            viewBox="0 0 20 20"
          >
            <path
              d="M6.707 5.293a1 1 0 0 0-1.414 1.414l3.293 3.293-3.293 3.293a1 1 0 1 0 1.414 1.414l3.293-3.293 3.293 3.293a1 1 0 0 0 1.414-1.414l-3.293-3.293 3.293-3.293a1 1 0 0 0-1.414-1.414l-3.293 3.293-3.293-3.293Z"
            />
          </svg>
        </span>
      </button>
    </span>
    <span
      className="Polaris-Tag Polaris-Tag--removable"
    >
      <span
        className="Polaris-Tag__TagText"
        title="tag2"
      >
        tag2
      </span>
      <button
        aria-label="Remove tag2"
        className="Polaris-Tag__Button"
        disabled={false}
        onClick={[Function]}
        onMouseUp={[Function]}
        type="button"
      >
        <span
          className="Polaris-Icon"
        >
          <span
            className="Polaris-Text--root Polaris-Text--visuallyHidden"
          />
          <svg
            aria-hidden="true"
            className="Polaris-Icon__Svg"
            focusable="false"
            viewBox="0 0 20 20"
          >
            <path
              d="M6.707 5.293a1 1 0 0 0-1.414 1.414l3.293 3.293-3.293 3.293a1 1 0 1 0 1.414 1.414l3.293-3.293 3.293 3.293a1 1 0 0 0 1.414-1.414l-3.293-3.293 3.293-3.293a1 1 0 0 0-1.414-1.414l-3.293 3.293-3.293-3.293Z"
            />
          </svg>
        </span>
      </button>
    </span>
    <span
      className="Polaris-Tag Polaris-Tag--removable"
    >
      <span
        className="Polaris-Tag__TagText"
        title="tag3"
      >
        tag3
      </span>
      <button
        aria-label="Remove tag3"
        className="Polaris-Tag__Button"
        disabled={false}
        onClick={[Function]}
        onMouseUp={[Function]}
        type="button"
      >
        <span
          className="Polaris-Icon"
        >
          <span
            className="Polaris-Text--root Polaris-Text--visuallyHidden"
          />
          <svg
            aria-hidden="true"
            className="Polaris-Icon__Svg"
            focusable="false"
            viewBox="0 0 20 20"
          >
            <path
              d="M6.707 5.293a1 1 0 0 0-1.414 1.414l3.293 3.293-3.293 3.293a1 1 0 1 0 1.414 1.414l3.293-3.293 3.293 3.293a1 1 0 0 0 1.414-1.414l-3.293-3.293 3.293-3.293a1 1 0 0 0-1.414-1.414l-3.293 3.293-3.293-3.293Z"
            />
          </svg>
        </span>
      </button>
    </span>
    <span
      className="sp"
    >
      <input
        className="input"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Tags"
        value=""
      />
    </span>
  </div>,
  <div
    id="PolarisPortalsContainer"
  />,
]
`;
